### ISSUE TITLE
Incorrect behaviour when heartbeat is missing

### ISSUE THREAD

**Issue by sicarrots** [php-amqplib/php-amqplib#787]
**Version(s) affected**: 2.11.1

**Description**  
When messages are published with basic_ack method after missing heartbeat and closed connection on rabbitmq site, basic_ack does not throws any exceptions. `PhpAmqpLib\Exception\AMQPHeartbeatMissedException` is thrown on channel->close().
From following reproducing code only "message 1" appears on queue "foo", messages 2-5 are missing though was published correctly. 

**How to reproduce**  
```php
<?php

require 'vendor/autoload.php';

use PhpAmqpLib\Connection\AMQPStreamConnection;
use PhpAmqpLib\Message\AMQPMessage;

const HEARTBEAT = 2;

$connection = new AMQPStreamConnection(
    'localhost',
    5672,
    'guest',
    'guest',
    '/',
    false,
    'AMQPLAIN',
    null,
    'en_US',
    20,
    20,
    null,
    false,
    HEARTBEAT
);
$channel = $connection->channel();

echo('publish 1' . PHP_EOL);
$channel->basic_publish(new AMQPMessage('message 1'), '', 'foo');
sleep(HEARTBEAT*3);
echo('publish 2' . PHP_EOL);
$channel->basic_publish(new AMQPMessage('message 2'), '', 'foo'); // Here i expected to exception be thrown
echo('publish 3' . PHP_EOL);
$channel->basic_publish(new AMQPMessage('message 3'), '', 'foo');
echo('publish 4' . PHP_EOL);
$channel->basic_publish(new AMQPMessage('message 4'), '', 'foo');
echo('publish 5' . PHP_EOL);
$channel->basic_publish(new AMQPMessage('message 5'), '', 'foo');
echo('end publishing' . PHP_EOL);
echo('close channel' . PHP_EOL);
$channel->close(); // here exception is thrown
echo('close connection' . PHP_EOL);
$connection->close();
```

**Additional context**  
MacOS 10.15.3, Docker 2.2.0.5, PHP 7.4.4, rabbitmq on docker: `docker run -i -d -p 5672:5672 -p 15672:15672 --name rabbit rabbitmq:3.7-management`

**Comment by lukebakken** [comment]
Thanks for this information. Could you also please do a TCP capture on port 5672 during this sequence of events? Compress the capture and attach it to this issue.

**Comment by sicarrots** [comment]
Capture using `sudo tcpdump port 5672 -i lo0 -vv -x -X -s 1500 -w capture` in attachment.
It seems that messages are delivered to rabbitmq but from some reason are not routed to queue after missing heartbeats. When I try to change `sleep(HEARTBEAT*3); ` to `sleep(HEARTBEAT);` messages are delivered properly.
[capture.tar.gz](https://github.com/php-amqplib/php-amqplib/files/4445121/capture.tar.gz)

**Comment by lukebakken** [comment]
Thanks. Finally, the RabbitMQ log file from this period of time would be helpful.

**Comment by sicarrots** [comment]
[capture.tar.gz](https://github.com/php-amqplib/php-amqplib/files/4445291/capture.tar.gz)
[rabbit.log](https://github.com/php-amqplib/php-amqplib/files/4445292/rabbit.log)
capture and rabbitmq log from the same time window

**Comment by lukebakken** [comment]
> It seems that messages are delivered to rabbitmq but from some reason are not routed to queue after missing heartbeats.

That's correct, RabbitMQ has already closed the connection and associated channel, so the messages are dropped.

**Comment by ramunasd** [comment]
@sicarrots here you use "normal" or so called ack-less mode, it means broker must not respond and you cannot be 100% certain about published messages.

Here is example how to enable and use "confirm" mode https://github.com/php-amqplib/php-amqplib/blob/master/demo/amqp_publisher_with_confirms.php

Related issue #493

**Comment by sicarrots** [comment]
@ramunasd I know that im using "normal" communication option, but in this particular case we known at library level that heartbeat is missing - library have to respond to ping request and we know which period of pings is used.
I've written this simple example in python using pika library, exception is thrown immediately when heartbeat is missing when trying to publish second message after sleep. Look at code and attached rabbitmq log and tcpdump.

```python
import pika
import time

connection = pika.BlockingConnection(pika.ConnectionParameters('localhost', heartbeat=2))
channel = connection.channel()

print('message 1')
channel.basic_publish(exchange='', routing_key='foo', body='message 1')
time.sleep(6)
print('message 2')
channel.basic_publish(exchange='', routing_key='foo', body='message 2') # here exception is thrown "pika.exceptions.StreamLostError: Transport indicated EOF"
print('message 3')
channel.basic_publish(exchange='', routing_key='foo', body='message 3')
print('message 4')
channel.basic_publish(exchange='', routing_key='foo', body='message 4')
print('message 5')
channel.basic_publish(exchange='', routing_key='foo', body='message 5')
print('close ch')
channel.close()
print('close conn')
connection.close()
print('end')
```
To run this example ensure you have python 3 and execute:
```bash
python -m venv .
. ./bin/activate
pip install pika
python test.py
```
[capture.tar.gz](https://github.com/php-amqplib/php-amqplib/files/4448897/capture.tar.gz)
[rabbit.log](https://github.com/php-amqplib/php-amqplib/files/4448898/rabbit.log)

**Comment by ramunasd** [comment]
We do have functional test for these situations, but it works only with bigger data frames. This is because we can get clear error only when internal OS buffers becomes full and real "data write" cannot be executed. Nonetheless, I will try to find solution at least for stream connections.

**Comment by ramunasd** [comment]
@sicarrots can You test latest master commit? If everything goes well we could release patch version 2.11.2

**Comment by sicarrots** [comment]
@ramunasd Yes, with example given in first comment now it works correctly. Thanks for fixing this!

**Comment by ramunasd** [comment]
@lukebakken the PR is added to next patch milestone. It's not too much for one version, but still very important fix and stability improvement.

**Comment by lukebakken** [comment]
OK, this issue was fixed by #791, correct? Is it time to release 2.11.2?

**Comment by ramunasd** [comment]
@lukebakken yes, it's correct, I'd like to release this fix as version 2.11.2

**Comment by lukebakken** [comment]
Fixed by #791

**Comment by thomasvargiu** [comment]
@ramunasd @lukebakken  Using the same library version I don't have the same beahviour, but on the second publish I had a `PhpAmqpLib\Exception\AMQPConnectionClosedException` exception.

Now with v1.11.2 there is this big bug: #793
